cty/value_init.go

```diff
@@ -186,6 +184,23 @@ func ListValEmpty(element Type) Value {
 	}
 }
 
+// CanListFunc returns an error if the given Values can not be coalesced
+// into a single List due to inconsistent element types.
+func CanListFunc(vals []Value) error {
```

Comment:
I think I created this confusion by making a typo when we were originally talking about this, but I was actually proposing for these to be named `CanListVal`, `CanMapVal`, etc, to represent that they each determine whether the corresponding `Can...` function can succeed.

Given that there's only one possible panic situation here, and for consistency with the other `Can...`-prefixed functions elsewhere in `cty`, I think I'd rather have these be functions that return `bool` to indicate that a subsequent call would succeed, rather than returning an error, and then let the `ListVal`/etc functions be responsible for constructing a suitable panic message as they were before.

This does mean that the `convert` codepaths will also need to construct themselves a suitable `error` value to return, but I think that's justified because the error messages there ought to be written with an end-user audience in mind, rather than a Go developer audience in mind. We don't really have enough context to return a great error message at that `convert` layer (cause we're already in a "shouldn't really happen" situation anyway), so I'd go for something pretty generic like this:

`List element types don't match`

A higher layer can replace that with a better error message if it has more interesting context to add, but we can save that for a later change.

Reply:
Got it! I reverted the corresponding constructor functions (so they maintain the element-specific error message) and refactored the `Can...` funcs to return bools (that was my original plan anyway, so cool!). I also added test coverage for the new functions.